File: python/src/simulate.py
"""Main I-V simulation driver.

Matches Julia FranckCondon.jl simulate_iv function and MATLAB run_benchmark.m
simulation loop exactly.
"""

import numpy as np
from .constants import ELEMENTARY_CHARGE
from .fc_matrix import FCCache
from .rate import RateStore, calculate_all_rateW
from .current import current_from_rate_equations


def simulate_iv(N, vmode, alphaL, alphaR, lambda_, Vsd_vec, T, eta, Vg, tau, verbose=True):
    """Run the full I-V simulation.

    Parameters
    ----------
    N : int
        Number of phonon states.
    vmode : float
        Phonon energy in eV.
    alphaL, alphaR : float
        Tunnel coupling ratios (Gamma/hbar*omega).
    lambda_ : float
        Electron-phonon coupling strength.
    Vsd_vec : ndarray
        Bias voltage sweep values.
    T : float
        Temperature in Kelvin.
    eta : float
        Voltage division factor.
    Vg : float
        Gate voltage.
    tau : float
        Phonon relaxation time.
    verbose : bool
        Print progress.

    Returns
    -------
    tuple of (ndarray, ndarray, ndarray, ndarray)
        (Vsd_vec, I_tol, I_seq, I_cot) in SI units (Amperes).
    """
    nVsd = len(Vsd_vec)
    I_tol = np.zeros(nVsd)
    I_seq = np.zeros(nVsd)
    I_cot = np.zeros(nVsd)

    # FC matrix cache persists across all bias points
    fc = FCCache(lambda_)

    for vv in range(nVsd):
        v = Vsd_vec[vv]
        if verbose:
            print(f"Vsd = {v:.4f} V  ({vv + 1}/{nVsd})")

        # Pre-compute rate matrices for both leads
        store = RateStore()
        calculate_all_rateW(store, N, vmode, alphaL, alphaR, lambda_, v, T, eta, 1, Vg, fc)
        calculate_all_rateW(store, N, vmode, alphaL, alphaR, lambda_, v, T, eta, -1, Vg, fc)

        # Compute current
        I_tol[vv], I_seq[vv], I_cot[vv] = current_from_rate_equations(
            v, N, vmode, alphaL, alphaR, lambda_, T, eta, Vg, tau, store)

        # Sign convention: current flows opposite to bias
        if v != 0.0:
            s = -np.sign(v)
            I_tol[vv] *= s
            I_seq[vv] *= s
            I_cot[vv] *= s

    # Convert to SI (Amperes)
    e = ELEMENTARY_CHARGE
    I_tol *= e
    I_seq *= e
    I_cot *= e

    return (Vsd_vec, I_tol, I_seq, I_cot)

File: fortran/src/simulate.f90
module fc_simulate
    use fc_constants, only: ELEMENTARY_CHARGE
    use fc_fc_matrix, only: fc_cache_t, fc_cache_init
    use fc_rate, only: rate_store_t, rate_store_init, rate_store_free, calculate_all_rateW
    use fc_current, only: current_from_rate_equations
    implicit none
contains

    subroutine simulate_iv(N, vmode, alphaL, alphaR, lambda, Vsd_vec, nVsd, &
                           T, eta, Vg, tau, verbose, &
                           I_tol_out, I_seq_out, I_cot_out)
        integer, intent(in) :: N, nVsd
        real(8), intent(in) :: vmode, alphaL, alphaR, lambda, T, eta, Vg, tau
        real(8), intent(in) :: Vsd_vec(nVsd)
        logical, intent(in) :: verbose
        real(8), intent(out) :: I_tol_out(nVsd), I_seq_out(nVsd), I_cot_out(nVsd)
        type(fc_cache_t) :: fc
        type(rate_store_t) :: store
        real(8) :: v, s_sign, cr_tol, cr_seq, cr_cot
        integer :: vv
        character(len=80) :: msg

        call fc_cache_init(fc, lambda)

        do vv = 1, nVsd
            v = Vsd_vec(vv)
            if (verbose) then
                write(msg, '(A,I4,A,I4,A,F7.4,A)') &
                    char(13)//'Bias point ', vv, '/', nVsd, ' (Vsd = ', v, ' V)'
                write(0, '(A)', advance='no') trim(msg)
            end if

            call rate_store_init(store, N)
            call calculate_all_rateW(store, N, vmode, alphaL, alphaR, lambda, &
                                      v, T, eta, 1, Vg, fc)
            call calculate_all_rateW(store, N, vmode, alphaL, alphaR, lambda, &
                                      v, T, eta, -1, Vg, fc)

            call current_from_rate_equations(v, N, vmode, alphaL, alphaR, lambda, &
                                             T, eta, Vg, tau, store, &
                                             cr_tol, cr_seq, cr_cot)

            if (v /= 0.0d0) then
                if (v > 0.0d0) then
                    s_sign = -1.0d0
                else
                    s_sign = 1.0d0
                end if
                cr_tol = cr_tol * s_sign
                cr_seq = cr_seq * s_sign
                cr_cot = cr_cot * s_sign
            end if

            I_tol_out(vv) = cr_tol * ELEMENTARY_CHARGE
            I_seq_out(vv) = cr_seq * ELEMENTARY_CHARGE
            I_cot_out(vv) = cr_cot * ELEMENTARY_CHARGE

            call rate_store_free(store)
        end do

        if (verbose) then
            write(0, *)
        end if
    end subroutine simulate_iv

end module fc_simulate
